ci/env/install-dependencies.sh

```diff
@@ -297,153 +297,163 @@ download_mnist() {
+        delayed_packages+=("--extra-index-url")
+        delayed_packages+=("https://pypi.fury.io/arrow-nightlies/")
+        delayed_packages+=("--prefer-binary")
+        delayed_packages+=("--pre")
+        delayed_packages+=("pyarrow")
       else
-        pip install -U pyarrow=="${ARROW_VERSION}"
+        delayed_packages+=("pyarrow==${ARROW_VERSION}")
       fi
     fi
     if [ -n "${ARROW_MONGO_VERSION-}" ]; then
-	pip install -U pymongoarrow=="${ARROW_MONGO_VERSION}"
+      delayed_packages+=("pymongoarrow==${ARROW_MONGO_VERSION}")
     fi
   fi
 
-  # Remove this entire section once Serve dependencies are fixed.
-  if [ "${MINIMAL_INSTALL-}" != 1 ] && [ "${DOC_TESTING-}" != 1 ] && [ "${TRAIN_TESTING-}" != 1 ] && [ "${TUNE_TESTING-}" != 1 ] && [ "${RLLIB_TESTING-}" != 1 ]; then
-    # If CI has deemed that a different version of Torch
-    # should be installed, then upgrade/downgrade to that specific version.
-    if [ -n "${TORCH_VERSION-}" ]; then
-      case "${TORCH_VERSION-1.9.0}" in
-        1.9.0) TORCHVISION_VERSION=0.10.0;;
-        1.8.1) TORCHVISION_VERSION=0.9.1;;
-        1.5) TORCHVISION_VERSION=0.6.0;;
-        *) TORCHVISION_VERSION=0.5.0;;
-      esac
-      pip install --use-deprecated=legacy-resolver --upgrade torch=="${TORCH_VERSION-1.9.0}" torchvision=="${TORCHVISION_VERSION}"
-    fi
+  if [ "${install_ml_no_deps}" = 1 ]; then
+    pip install -r "${WORKSPACE_DIR}/python/requirements/ml/requirements_no_deps.txt"
+  fi
+
+  retry_pip_install "CC=gcc pip install -Ur ${WORKSPACE_DIR}/python/requirements.txt"
+
+  # Install deeplearning libraries (Torch + TensorFlow)
+  if [ -n "${TORCH_VERSION-}" ] || [ "${DL-}" = "1" ] || [ "${RLLIB_TESTING-}" = 1 ] || [ "${TRAIN_TESTING-}" = 1 ] || [ "${TUNE_TESTING-}" = 1 ]; then
+      # If we require a custom torch version, use that
+      if [ -n "${TORCH_VERSION-}" ]; then
+        case "${TORCH_VERSION-1.9.0}" in
+          1.9.0) TORCHVISION_VERSION=0.10.0;;
+          1.8.1) TORCHVISION_VERSION=0.9.1;;
+          1.6) TORCHVISION_VERSION=0.7.0;;
+          1.5) TORCHVISION_VERSION=0.6.0;;
+          *) TORCHVISION_VERSION=0.5.0;;
+        esac
+        # Install right away, as some dependencies (e.g. torch-spline-conv) need
+        # torch to be installed for their own install.
+        pip install -U "torch==${TORCH_VERSION-1.9.0}" "torchvision==${TORCHVISION_VERSION}"
+        # We won't add requirements_dl.txt as it would otherwise overwrite our custom
+        # torch. Thus we have also have to install tensorflow manually.
+        # Keep it sync with requirements_dl.txt
+        pip install -U "tensorflow==2.11.0" "tensorflow-probability==0.19.0"
+      else
+        # Otherwise, use pinned default torch version.
+        # Again, install right away, as some dependencies (e.g. torch-spline-conv) need
+        # torch to be installed for their own install.
+        # Keep it sync with requirements_dl.txt
+        pip install torch==1.13.0 torchvision==0.14.0
```

Comment:
this can easily get out of sync with `requirements_dl.txt`. How come in current master we don't need to do this in a separate step and can do everything in the same file?

If having them separate is a necessity, then I would propose we have 2 different requirements files, rather than burying the torch install in this script.